kotlinx-coroutines-core/common/src/internal/SegmentQueueSynchronizer.kt

```diff
@@ -0,0 +1,606 @@
+import kotlinx.coroutines.internal.SegmentQueueSynchronizer.ResumeMode.SYNC
+import kotlinx.coroutines.internal.SegmentQueueSynchronizer.CancellationMode.*
+import kotlinx.coroutines.sync.*
+import kotlin.coroutines.*
+import kotlin.native.concurrent.*
+
+/**
+ * [SegmentQueueSynchronizer] is an abstraction for implementing _fair_ synchronization
+ * and communication primitives. It maintains a FIFO queue of waiting requests and is
+ * provided with two main functions:
+ * + [suspend] that stores the specified waiter into the queue, and
+ * + [resume] that tries to retrieve and resume the first waiter with the specified value.
+ *
+ * One may consider this abstraction as an infinite array with two counters that reference the next cells
+ * for enqueueing a continuation in [suspend] and for retrieving it in [resume]. To be short, when
+ * [suspend] is invoked, it increments the corresponding counter via fast `Fetch-And-Add` and stores the
+ * continuation into the cell. At the same time, [resume] increments its own counter and comes to the
+ * corresponding cell.
+ *
+ * A typical implementation via [SegmentQueueSynchronizer] performs some synchronization at first,
+ * (e.g., [Semaphore] modifies the number of available permits), and invokes [suspend] or [resume]
+ * after that. Following this pattern, it is possible in a concurrent environment that [resume]
+ * is executed before [suspend] (similarly to the race between `park` and `unpark` for threads),
+ * so that [resume] comes to a cell in the empty state. This race can be solved with two [strategies][ResumeMode]:
+ * [asynchronous][ASYNC] and [synchronous][SYNC]. In the [asynchronous][ASYNC] mode, [resume] puts
+ * the value into the cell if it is still empty and finishes immediately; this way, the following [suspend] comes
+ * to this cell and simply grabs the element without suspension. At the same time, in the [synchronous][SYNC] mode,
+ * [resume] waits in a bounded spin-loop cycle until the put element is taken by a concurrent [suspend],
+ * marking the cell as broken if it is not taken after the spin-loop is finished. In this case, both the current
+ * [resume] and the [suspend] that comes to this broken cell fail and the corresponding operations on the data
+ * structure are typically restarted from the beginning.
+ *
+ * Since [suspend] can store [CancellableContinuation]-s, it is possible for [resume] to fail if the
+ * continuation is already cancelled. In this case, most of the algorithms retry the whole operation.
+ * However, if there are many consecutive cancelled waiters, it seems more efficient to skip them
+ * somehow. Thus, there are different [cancellation modes][cancellationMode] that can be used.
+ * In the simple mode ([SIMPLE]), which is used by default, [resume] simply fails on cancelled waiters as
+ * described above, while in smart modes ([SMART_SYNC] and [SMART_ASYNC]) [resume] skips cells in the
```

Comment:
The specific semantics of skipping are fuzzy here. If a cell is cancelled in a segment that is not removed, for example, this cell is not skipped, so saying that [resume] skips cells is not strictly correct. Another difference in the implementation is that [resume] does not fail when encountering a cancelled cell but proceeds to wake another one up, but this is not "skipping", as the cancelled cell is still visited, but more like guaranteeing that each call to [resume] wakes at least one waiter up if it's possible at all.

Reply:
We do not have segments in the abstraction, cells in the CANCELLED state are always skipped when a smart cancellation mode is used.